ofrak_core/ofrak/core/free_space.py

```diff
@@ -114,12 +114,16 @@ async def allocate(
     async def allocate_bom(
         self,
         bom: BOM,
+        permission_map: Optional[Mapping[MemoryPermissions, Iterable[MemoryPermissions]]] = None,
```

Comment:
Do we need the full flexibility that this map allows? For example, allowing executable segments to be allocated to (non-executable) read/write space. There seem to be only two legit fallback cases:

* R -> RW
* R -> RX

It would be less flexible, but maybe usage would be more clear and the options more relevant if we just had two optional flags, `allow_r_in_rw` and `allow_r_in_rx` (or similar self-documenting names).

Reply:
I go back and forth.

If someone asked for the ability to map their patch RX into their destination's RW would we say no? 

(considering beyond Linux platforms or ELF formats...)

Reply:
I guess we enable them to do that with your bools.

Reply:
Yeah, the flexibility to deal with unusual architectures or even buggy toolchains is a big bonus. Also being able to set a prioritization of RX over RW or vice versa. I think these come at the disadvantage of the interface being a little harder to grok, so there's the risk that users won't find this option, even when it would be useful. But I don't think that has to block this PR.

Another alternative would be adding 3 new options instead of one: `r_overflow_to`, `rx_overflow_to`, and `rw_overflow_to`. Each one would be an `Iterable[MemoryPermissions]`, essentially unfolding the `permission` map into the function arguments, defaulting to `None` or an empty Tuple.  I think this will be a little easier to understand for a user.

Again, that doesn't have to block this PR. Everything else looks good. I'll leave this comment, if you like the idea, run with it. Otherwise I'll approve and you can merge at your convenience.

Reply:
IMHO `permission_map` seems like the easiest option to me -- it creates only one optional argument. If a user runs into a permission mapping problem, I think they are in a situation where they can figure out how to use it.

I'd vote for leaving it in this PR and returning to this if it causes issues down the line.

Reply:
What Wyatt is outlines here is the original reason for why we've done it this way. Let's go with our first impression of what is "simple" here (fewer parameters, more possible expressiveness), and if we run into to trouble we can adjust to something like Ed's suggestion.